### PR TITLE
[RFC] Push to create

### LINKED ISSUE #376: Scripted Repository Creation
It would be wonderful if we could create repositories from command line in order to automate some processes !!

### PR DESCRIPTION
**Pull request by willdean**
Fixes #376.  Was started in PR #460 

@RedX2501 I made quite a big change to the Integration tests to provide a GitResult class rather than using the tuple (all the Item1,Item2,Item3 stuff confuses me).   I hope this isn't offensive to you - if it causes you hassle with changes you're making in parallel, I'm happy to do any tedious merge work which arises.

I feel happier in tests I write to see the asserted conditions more explicitly in the test method rather implied in the helper methods, so I've tried to allow the result to be returned to the test method instead of being consumed in the helper, but then added code in the result object to make testing stuff easy.   I would value your opinion.

I cannot get the two tests for push-to-create to work here on my machine - using the admin:admin@ style of URL with the server instance which SpecsForMVC spins-up always suffers an auth error (the client never sends the HTTP basic auth header).    I don't know why - I can use the same kind URL with the same code if I run it in Visual Studio, and the admin/admin account is OK for web-login into the SpecForMVC server instance.   However, they do work OK on Appveyer.

This PR also has a more robust DirectoryDelete routine in the integration tests, which may help with intermittent test failures.

### CODE REVIEW

File: Bonobo.Git.Server.Test/IntegrationTests/GitResult.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                Assert.Fail("Git operation failed with exit code {0}, stderr {1}", ExitCode, StdErr);
+            }
+        }
+
+        public void ErrorMustMatch(MsysgitResources.Definition resource, params object[] args)
+        {
+            string matchString;
+            if (args.Length > 0)
+            {
+                matchString = string.Format(Resources[resource], args);
```
Comment (RedX2501) [inline Bonobo.Git.Server.Test/IntegrationTests/GitResult.cs]: Will you get an error if args.length == 0? Else i'd remove the branch as it is a pointless optimization IMHO.

File: Bonobo.Git.Server.Test/IntegrationTests/GitResult.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            }
+            else
+            {
+                matchString = Resources[resource];
+            }
+            var expected = matchString.Trim();
+            var actual = StdErr.Trim();
+            if (expected != actual)
+            {
+                Assert.Fail("Git operation StdErr mismatch - expected '{0}', was '{1}'", expected, actual);
```
Comment (RedX2501) [inline Bonobo.Git.Server.Test/IntegrationTests/GitResult.cs]: The format for all other Assert comparisons is actual, expected. Maybe we should also do it here for consistency. Why not use Assert.AreEqual here anyway?

File: Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs [hunk 3]
```diff
@@ -243,8 +246,45 @@ public void NamedPushToAnonRepo()
...
+                CreateAndAddFiles(git);
+
+                RunGitOnRepo(git, "push origin master").ErrorMustMatch(MsysgitResources.Definition.RepositoryNotFoundError, RepositoryUrlWithCredentials);
+            });
+        }
+
+        [TestMethod, TestCategory(TestCategories.ClAndWebIntegrationTest)]
+        public void PushToCreateIsAllowedIfOptionIsSet()
+        {
+            ForAllGits((git, resource) =>
```
Comment (RedX2501) [inline Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs]: I thought you wanted to use your new Git object here instead of the tuple? Did i miss something?

File: Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs [hunk 4]
```diff
@@ -378,15 +423,31 @@ private void InitAndPullRepository(string git, MsysgitResources resource)
...
+
+        private void InitRepository(string git)
+        {
+            RunGitOnRepo(git, "init").ExpectSuccess();
+            RunGitOnRepo(git, String.Format("remote add origin {0}", RepositoryUrlWithCredentials)).ExpectSuccess();
+        }
+		
+        private void InitAndPushRepository(string git, MsysgitResources resource)
+        {
+            RunGitOnRepo(git, "init");
```
Comment (RedX2501) [inline Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs]: You should alse check for success in all these calls to early detect any problems.

File: Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs [hunk 5]
```diff
@@ -378,15 +423,31 @@ private void InitAndPullRepository(string git, MsysgitResources resource)
...
             var result = RunGitOnRepo(git, "pull origin master");
 
-            Assert.AreEqual(String.Format(resource[MsysgitResources.Definition.PullRepositoryError], RepositoryUrlWithoutCredentials), result.Item2);
+            Assert.AreEqual(String.Format(resource[MsysgitResources.Definition.PullRepositoryError], RepositoryUrlWithoutCredentials), result.StdErr);
+        }
+
+
+        private void InitRepository(string git)
+        {
+            RunGitOnRepo(git, "init").ExpectSuccess();
```
Comment (RedX2501) [inline Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs]: Are you sure that only checking the exit code is enough?

File: Bonobo.Git.Server/App_GlobalResources/Resources.resx [hunk 6]
```diff
@@ -682,6 +682,12 @@
   <data name="Settings_Global_AllowUserRepositoryCreation_Hint" xml:space="preserve">
     <value>If checkbox is unchecked, only administrators can create repositories.</value>
   </data>
+  <data name="Settings_Global_AllowPushToCreate" xml:space="preserve">
```
Comment (RedX2501) [inline Bonobo.Git.Server/App_GlobalResources/Resources.resx]: I think that as all other repo configuration stuff, there should be a repo local setting for it also. Or even only a repo local setting. If you wish i can create the update scripts for it.

File: Bonobo.Git.Server/Controllers/GitController.cs [hunk 7]
```diff
@@ -80,6 +102,36 @@ public ActionResult SecureReceivePack(String repositoryName)
             }
         }
 
+        private bool TryCreateOnPush(string repositoryName)
+        {
+            DirectoryInfo directory = GetDirectoryInfo(repositoryName);
```
Comment (RedX2501) [inline Bonobo.Git.Server/Controllers/GitController.cs]: Don't you have to combine this with the current setting for repositorypath?

File: Bonobo.Git.Server.Test/IntegrationTests/GitResult.cs (hunk 1 above)
Comment (willdean) [inline Bonobo.Git.Server.Test/IntegrationTests/GitResult.cs]: It's not really an opptimisation, if you don't have any args then we don't try to expand the format string, sonyou can inclue curly brackets without a problem.  But you're probably right it's unnecessary.

File: Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs (hunk 3 above)
Comment (willdean) [inline Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs]: I didn't do that mod, I wasn't sure what you thought about it

File: Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs (hunk 4 above)
Comment (willdean) [inline Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs]: Yes

File: Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs (hunk 5 above)
Comment (willdean) [inline Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs]: I think it's a reliable check for success, when it's not the main assertion of the test method.

File: Bonobo.Git.Server/App_GlobalResources/Resources.resx (hunk 6 above)
Comment (willdean) [inline Bonobo.Git.Server/App_GlobalResources/Resources.resx]: If you can write a per-repo setting for permission to create the repo which contains the permission option, then be my guest :-).   Funny enough i went down the same line of thought before realising the problem.

File: Bonobo.Git.Server/Controllers/GitController.cs (hunk 7 above)
Comment (willdean) [inline Bonobo.Git.Server/Controllers/GitController.cs]: Yup this is wrong, thanks

**Comment by willdean** [comment]
Thanks for the feedback, will have a goes at this tomorrow

**Comment by RedX2501** [comment]
I noticed that we have not being updated the changelog file properly. Do you mind adding an entry for this PR?

File: Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs (hunk 3 above)
Comment (RedX2501) [inline Bonobo.Git.Server.Test/IntegrationTests/MsysgitIntegrationTests.cs]: Oh sorry. I'm totally fine with any change that will increase separation of concerns, increases code readability and reduces code duplication. We should only try to keep helper methods in the IntegrationTestHelpers (maybe they only exist in my branch yet) so that we can reuse them in the not Git tests.

Maybe we should make an own PR for it.

File: Bonobo.Git.Server/Controllers/GitController.cs (hunk 7 above)
Comment (willdean) [inline Bonobo.Git.Server/Controllers/GitController.cs]: Oh no, it's correct actually - GetDirectoryInfo does this.
